### PR TITLE
optbuilder: refactor semantic analysis of FOR UPDATE

### PR DESCRIPTION
**Pull request by michae2**
**optbuilder: a few minor tweaks to building of FOR UPDATE**

Make a few minor tweaks to semantic analysis of FOR UPDATE locking
clauses.

1. Disallow multiple FOR UPDATE clauses on parenthesized queries. We do
   not currently handle scopes of parenthesized queries correctly.
   Because of this, we disallow multiple ORDER BY and LIMIT clauses on
   parenthesized queries. The previous implementation of FOR UPDATE was
   simple enough that we could work around this, but the upcoming
   changes make it impossible to support.

2. Allow FOR UPDATE on statements with VALUES in the FROM list (but
   continue to disallow FOR UPDATE on VALUES directly). This matches
   Postgres.

Release note (sql change): Two minor changes to FOR UPDATE clauses:

1. Multiple FOR UPDATE clauses on fully-parenthesized queries are now
disallowed. For example, the following statements are now disallowed:

```
(SELECT 1 FOR UPDATE) FOR UPDATE;
SELECT * FROM ((SELECT 1 FOR UPDATE) FOR UPDATE) AS x;
```

whereas statements like the following are still allowed:

```
SELECT * FROM (SELECT 1 FOR UPDATE) AS x FOR UPDATE;
SELECT (SELECT 1 FOR UPDATE) FOR UPDATE;
```

This does not match PostgreSQL, which allows all of these, but does
match our behavior for ORDER BY and LIMIT.

2. FOR UPDATE is now allowed on statements with VALUES in the FROM list
or as a subquery. For example, the following statements are now allowed:

```
SELECT (VALUES (1)) FOR UPDATE;
SELECT * FROM (VALUES (1)) AS x FOR UPDATE;
```

Using FOR UPDATE directly on VALUES is still disallowed:

```
VALUES (1) FOR UPDATE;
(VALUES (1)) FOR UPDATE;
INSERT INTO t VALUES (1) FOR UPDATE;
```

This matches PostgreSQL.

**optbuilder: refactor semantic analysis of FOR UPDATE**

Locking clauses such as FOR UPDATE and FOR SHARE apply to some or all of
the data sources in a query's FROM list, depending on whether they have
targets (FOR UPDATE OF x). Without targets, they always apply within
subqueries in the FROM list. With targets, they apply within subqueries
if the subquery alias matches the target.

Because of this scope-like nature of FOR UPDATE and FOR SHARE, we
implement semantic analysis using a stack of locking items that grow as
we build inner subqueries deeper in the recursive optbuilder calls.

Prior to this change, we only used the stack of locking items during
buildScan, at the very bottom of the recursion. Because of this, calls
to `lockingSpec.filter` could afford to compress the stack into a single
locking item on our way deeper in the recursion.

As part of the upcoming fix for https://github.com/cockroachdb/cockroach/issues/75457, however, we will need to build a
new Lock operator when popping off locking items after returning from
the recursion. That Lock operator will need some information gathered
from buildScan at the bottom of the recursion.

To support this, we refactor the stack of locking items to be two
stacks: one that tracks all locking items in scope, and a second that
tracks which locking items currently apply. This will allow buildScan to
associate table information with the correct locking item(s), which can
then be used to build Lock operators when popping the locking items.

As a bonus, by using only the applied locking item stack in
`validateLockingInFrom` we can make validation of SELECT FOR UPDATE
queries a little more precise, which allows some queries we were
incorrectly disallowing.

Informs: https://github.com/cockroachdb/cockroach/issues/57031, https://github.com/cockroachdb/cockroach/issues/75457

Epic: CRDB-25322

Release note (sql change): Allow FOR UPDATE on some queries that were
previously disallowed. Queries that use the following operations are now
allowed to have FOR UPDATE OF as long as the prohibited operation is in
a subquery not locked by the FOR UPDATE OF:

- UNION
- INTERSECT
- EXCEPT
- DISTINCT
- GROUP BY
- HAVING
- aggregations
- window functions

For example, the following query is now allowed because the subquery
using the prohibited operations is not affected by the FOR UPDATE OF:

```
SELECT * FROM t,
  (SELECT DISTINCT 0, 0 UNION SELECT a, count(*) FROM t GROUP BY a HAVING a > 0) AS u
  FOR UPDATE OF t;
```

This matches PostgreSQL.

### CODE REVIEW

**Comment by michae2** [comment]
TLDR: this changes

```go
type lockingSpec []*tree.LockingItem
```

to

```go
type lockingSpec []*lockingItem
```

And then replaces almost all uses of `lockingSpec` in select.go with `lockingContext`, which is:

```go
type lockingContext struct {
  lockScope []*lockingItem
  locking lockingSpec
}
```

**Comment by michae2** [comment]
The test failures have been flakes. This is RFAL.

**Review by DrewKimball — APPROVED** [review]
:lgtm: Great work! This is very well commented. I just have some test suggestions and a question.

> Reviewed 10 of 10 files at r3, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-:-Nfu0xBE8-N98kIHv-2o:bnsbjah)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, @msirek, and @nvanbenschoten)

___
*[`-- commits` line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-NfqlmZT4mbSyAYwvqnd):*
Do we have tests somewhere for allowed cases like these?
___
*[`pkg/sql/opt/optbuilder/locking.go` line 160 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-Nftxl205xcUpg-WvXTL:-Nftxl205xcUpg-WvXTM:b6z26pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/opt/optbuilder/locking.go#L160)):*
> ```Go
> 
> // blankLockingScope is a sentinel locking item that, when pushed, prevents
> // lockCtx.filter from matching targets outside it.
> ```

Is "it" the current scope?
___
*[`pkg/sql/opt/optbuilder/select.go` line 1148 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-Nfu0F47A8SD9JD1rK6a:-Nfu0F47A8SD9JD1rK6b:b-nh6fe2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/opt/optbuilder/select.go#L1148)):*
> ```Go
> 	}
> 
> 	for range lockingClause {
> ```

Huh, I didn't know this was possible.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_update` line 168 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-Nfu0abo2kRIqIwMKEzx:-Nfu0abo2kRIqIwMKEzy:bv5pofb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/logictest/testdata/logic_test/select_for_update#L168)):*
> ```
> 
> query II
> SELECT * FROM (SELECT 1 UNION SELECT 1) a, i FOR UPDATE OF i
> ```

Could we do another test with `FOR UPDATE OF a`?

**Review by nvb — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, 2 of 10 files at r2, 10 of 10 files at r3, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111258)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @msirek)

___
*[`-- commits` line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-NfY2QLIAhaJTv77wdF2):*
These release notes are great.
___
*[`pkg/sql/opt/optbuilder/select.go` line 1100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-NfY1LRo7NbaRWcWt4NC:-NfY1LRo7NbaRWcWt4ND:bm9cokh) ([raw file](https://github.com/cockroachdb/cockroach/blob/efdd0371e1ecd504f1f15914c156a0011ceff014/pkg/sql/opt/optbuilder/select.go#L1100)):*
> ```Go
> 	inScope *scope,
> ) (outScope *scope) {
> 	locking.apply(lockingClause)
> ```

`lockingSpec.apply` modifies the receiver, not the parameter. We're then passing the parameter to `rejectIfLocking` for VALUES clauses, but still passing the receiver for `UNION` clauses. That's intentional?

If so, consider a comment explaining why we aren't passing the entire lockingSpec down on the VALUES paths.

EDIT: this is different in the second commit. I believe it no longer applies.
___
*[`pkg/sql/opt/optbuilder/select.go` line 1148 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-Nfvgh4e72_37CMVkeH9:-Nfvgh4e72_37CMVkeHA:b-7z4ew6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/opt/optbuilder/select.go#L1148)):*
> ```Go
> 	}
> 
> 	for range lockingClause {
> ```

Consider giving this a comment, explaining the two different roles that it is playing.

**Review by msirek — APPROVED** [review]
Very nice work!

> Reviewed 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-:-NfwnCPM862FUUNgb3KS:bdm1mhe)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/optbuilder/locking.go` line 122 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-NfwohHX8tUP2Hc9PSC2:-NfwohHX8tUP2Hc9PSC3:b-x4343e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/opt/optbuilder/locking.go#L122)):*
> ```Go
> 	// might include locking items that do not currently apply because they have
> 	// an unmatched target.
> 	lockScope []*lockingItem
> ```

Is there a reason `lockScope` is a slice, while `locking` is a type with the same slice definition?  Is it just to indicate `locking` is the `Spec` of actually applicable items? Just wondering if `lockingSpec` should be used in both cases.

**Review by michae2 — COMMENTED** [review]
TFTRs! I'll go ahead and bors this and then if there are more comments, I can make changes in the next PR.

bors r=DrewKimball,msirek

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-:-Ng2qaG8Baeha5_8tTrZ:bih2hko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, @msirek, and @nvanbenschoten)

___
*[`-- commits` line 32 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-NfqlmZT4mbSyAYwvqnd):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Do we have tests somewhere for allowed cases like these?
</blockquote></details>

Oh, good call. I added these to `pkg/sql/logictest/testdata/logic_test/select_for_update`.
___
*[`pkg/sql/opt/optbuilder/locking.go` line 122 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-NfwohHX8tUP2Hc9PSC2:-Ng2kVbu6wyUjEkoIdhD:bcpsls) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/opt/optbuilder/locking.go#L122)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Is there a reason `lockScope` is a slice, while `locking` is a type with the same slice definition?  Is it just to indicate `locking` is the `Spec` of actually applicable items? Just wondering if `lockingSpec` should be used in both cases.
</blockquote></details>

Good question. Yes, it is to differentiate between the applied items and the items that are in scope but not necessarily applied. For example, it wouldn't make sense to call `.isSet()` or `.get()` on the items that are in scope but not necessarily applied.

I also did it to make the diff a little more meaningful (`locking lockingSpec` has the same purpose both before and after this PR: to describe the currently-applied locking).
___
*[`pkg/sql/opt/optbuilder/locking.go` line 160 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-Nftxl205xcUpg-WvXTL:-Ng2ln629NGjph05Vbc7:bijcxy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/opt/optbuilder/locking.go#L160)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Is "it" the current scope?
</blockquote></details>

"It" is the blank locking scope that was pushed.

For example, if we have a statement like:

```sql
SELECT * FROM t, (SELECT * FROM s, t FOR UPDATE OF s FOR SHARE) AS u FOR UPDATE OF t;
```

Then when we're building the FROM clause of the subquery the `lockCtx.lockScope` will look like:

```
[ { strength: FOR UPDATE, targets: [t] }, blankLockingScope, { strength: FOR SHARE, targets: [] }, { strength: FOR UPDATE, targets: [s] } ]
```

And inside the subquery, only the FOR SHARE locking applies to `t`.
___
*[`pkg/sql/opt/optbuilder/select.go` line 1100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-NfY1LRo7NbaRWcWt4NC:-Ng2m1NLBtCcCQJpimNG:b300oc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/efdd0371e1ecd504f1f15914c156a0011ceff014/pkg/sql/opt/optbuilder/select.go#L1100)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`lockingSpec.apply` modifies the receiver, not the parameter. We're then passing the parameter to `rejectIfLocking` for VALUES clauses, but still passing the receiver for `UNION` clauses. That's intentional?

If so, consider a comment explaining why we aren't passing the entire lockingSpec down on the VALUES paths.

EDIT: this is different in the second commit. I believe it no longer applies.
</blockquote></details>

This was intentional, but yes it is confusing. Added a comment. For `VALUES` we only check the immediate statement for locking, whereas for `UNION` we check all outer statements as well. It is to match Postgres.
___
*[`pkg/sql/opt/optbuilder/select.go` line 1148 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-Nfu0F47A8SD9JD1rK6a:-Ng2mMi_15mNi2sNXIuO:b3gn6oi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/opt/optbuilder/select.go#L1148)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Huh, I didn't know this was possible.
</blockquote></details>

Me neither! `crlfmt` did this for me!
___
*[`pkg/sql/opt/optbuilder/select.go` line 1148 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-Nfvgh4e72_37CMVkeH9:-Ng2mQj4EyU7A7FrGhlW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/opt/optbuilder/select.go#L1148)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider giving this a comment, explaining the two different roles that it is playing.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/select_for_update` line 168 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/111258#-Nfu0abo2kRIqIwMKEzx:-Ng2ne_ADNRVBcWzpOv9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9645995da46467acf7b5105a331c9f81313cb06/pkg/sql/logictest/testdata/logic_test/select_for_update#L168)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could we do another test with `FOR UPDATE OF a`?
</blockquote></details>

Done.
